### PR TITLE
Limitation of project resource

### LINKED ISSUE #27: Unable to assign large number of existing repos to project
**Describe the bug**
When assigning a large number (200+) existing repos to project(s) using the `repos` attribute, some assignments fail with 400 error. The error message suggests Artifactory isn't able to service all the concurrent API calls.

**Expected behavior**
All the repos be assigned to the respective projects

**Additional context**

See https://github.com/jfrog/terraform-provider-artifactory/issues/329#issuecomment-1056898960

### PR DESCRIPTION
**Pull request by dasmanas**
closes #27 with limitations
Supported for default 100 repos to be assigned to a project
request retry mechanism introduced on specific error conditions

- **repos** (Set of String) List of existing repo keys to be assigned to the project. You can set a maximum 100 repo keys. (se [below for limitations on the maximum number of repos](#repoNumLimitations)). You can override the default value by setting an environment value `REPO_LIMIT_OVERRIDE` to `true`. The default value of `REPO_LIMIT_OVERRIDE` is `false` when not set. E.g. To override the default maximum allowable elements of `repos` attribute use `export REPO_LIMIT_OVERRIDE=true`.

### CODE REVIEW

File: docs/resources/project.md [hunk 1]
```diff
@@ -84,7 +84,7 @@ resource "project" "myproject" {
 - **id** (String) The ID of this resource.
 - **max_storage_in_gibibytes** (Number) Storage quota in GiB. Must be 1 or larger. Set to -1 for unlimited storage. This is translated to binary bytes for Artifactory API. So for 1TB quota, this should be set to 1024 (vs 1000) which will translate to 1099511627776 bytes for the API.
 - **member** (Block Set) Member of the project. Element has one to one mapping with the [JFrog Project Users API](https://www.jfrog.com/confluence/display/JFROG/Artifactory+REST+API#ArtifactoryRESTAPI-UpdateUserinProject). (see [below for nested schema](#nestedblock--member))
-- **repos** (Set of String) List of existing repo keys to be assigned to the project.
+- **repos** (Set of String) List of existing repo keys to be assigned to the project. You can set a maximum 100 repo keys. (se [below for limitations on the maximum number of repos](#repoNumLimitations)). You can override the default value by setting an environment value `REPO_LIMIT_OVERRIDE` to `true`. The default value of `REPO_LIMIT_OVERRIDE` is `false` when not set. E.g. To override the default maximum allowable elements of `repos` attribute use `export REPO_LIMIT_OVERRIDE=true`.
```
Comment (chb0github) [inline docs/resources/project.md:87]: You have a typo here

`se[`

Comment (chb0github) [reply docs/resources/project.md:87]: And you're using square brackets. What formatting are you trying to achieve

File: pkg/project/util.go [hunk 2]
```diff
@@ -175,3 +181,27 @@ var intersection = apply(func(bs []Equatable, a Equatable) bool {
...
 })
+
+var getBoolEnvVar = func(key string, fallback bool) (bool, error) {
+	value, exists := os.LookupEnv(key)
+	if exists {
+		boolValue, err := strconv.ParseBool(value)
+		if err == nil {
+			return boolValue, nil
+		} else {
+			return fallback, fmt.Errorf("WARN: environment value of %s should be of type Boolean. Setting to a fallback value %t", key, fallback)
```
Comment (chb0github) [inline pkg/project/util.go:192]: Putting the word `WARN` doesn't make it any less of an error to go.

File: pkg/project/util.go [hunk 3]
```diff
@@ -175,3 +181,27 @@ var intersection = apply(func(bs []Equatable, a Equatable) bool {
 var difference = apply(func(bs []Equatable, a Equatable) bool {
 	return !contains(bs, a)
 })
+
+var getBoolEnvVar = func(key string, fallback bool) (bool, error) {
+	value, exists := os.LookupEnv(key)
+	if exists {
+		boolValue, err := strconv.ParseBool(value)
+		if err == nil {
+			return boolValue, nil
```
Comment (chb0github) [inline pkg/project/util.go:190]: You could summarize this whole function yo the point of not needing it. When i get back to my laptop I'll show you.

File: docs/resources/project.md (hunk 1 above)
Comment (alexhung) [reply docs/resources/project.md:87]: The description does not explain clearly how that env var affects the limitation of 100 repo keys.

File: pkg/project/resource_project_repo_test.go [hunk 4]
```diff
@@ -114,3 +116,131 @@ func TestAccProjectRepo(t *testing.T) {
...
+		}
+		return repoNames
+	}(numRepos)
+
+	repoNamesStr := func(repoNames []string) string {
+		jsonStr, err := json.Marshal(repoNames)
+		if err != nil {
+			return "[]"
+		}
+		return string(jsonStr)
```
Comment (alexhung) [inline pkg/project/resource_project_repo_test.go:145]: Why the need to convert to string again?

File: pkg/project/resource_project_repo_test.go [hunk 5]
```diff
@@ -114,3 +116,131 @@ func TestAccProjectRepo(t *testing.T) {
...
+		ProviderFactories: testAccProviders(),
+		Steps: []resource.TestStep{
+			{
+				Config: initialConfig,
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(resourceName, "key", fmt.Sprintf("%s", params["project_key"])),
+					resource.TestCheckResourceAttr(resourceName, "description", "test description"),
+					resource.TestCheckNoResourceAttr(resourceName, "repos"),
+					resource.TestCheckResourceAttr(resourceName, "repos.#", "0"),
+					//resource.TestCheckResourceAttr(resourceName, "repos.0", repo1),
```
Comment (alexhung) [inline pkg/project/resource_project_repo_test.go:221]: Please clean up commented out code.

File: pkg/project/resource_project_repo_test.go [hunk 6]
```diff
@@ -114,3 +116,131 @@ func TestAccProjectRepo(t *testing.T) {
...
+			},
+			{
+				Config: addRepoConfig,
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(resourceName, "key", fmt.Sprintf("%s", params["project_key"])),
+					resource.TestCheckResourceAttr(resourceName, "display_name", name),
+					resource.TestCheckResourceAttr(resourceName, "description", "test description"),
+					resource.TestCheckResourceAttr(resourceName, "repos.#", strconv.Itoa(numRepos)),
+					//resource.TestCheckResourceAttr(resourceName, "repos.0", repo1),
+					//resource.TestCheckResourceAttr(resourceName, "repos.1", repo2),
```
Comment (alexhung) [inline pkg/project/resource_project_repo_test.go:232]: Same here

File: pkg/project/util.go (hunk 2 above)
Comment (alexhung) [reply pkg/project/util.go:192]: Why even bother with returning `error` when https://github.com/jfrog/terraform-provider-project/pull/32/files#diff-78bdb51875e934e7c55ab887da3d8d2afcd35bc21d4684815ead7cf26392c844R228 ignores the error.

File: pkg/project/resource_project_repo_test.go (hunk 4 above)
Comment (dasmanas) [reply pkg/project/resource_project_repo_test.go:145]: jsonStr is of type []byte

File: pkg/project/resource_project_repo_test.go (hunk 5 above)
Comment (dasmanas) [reply pkg/project/resource_project_repo_test.go:221]: done

File: pkg/project/resource_project_repo_test.go (hunk 6 above)
Comment (dasmanas) [reply pkg/project/resource_project_repo_test.go:232]: done

File: pkg/project/util.go (hunk 2 above)
Comment (dasmanas) [reply pkg/project/util.go:192]: removed the error return as we are not going to throw error to CLI

File: docs/resources/project.md (hunk 1 above)
Comment (dasmanas) [reply docs/resources/project.md]: updated

File: docs/resources/project.md [hunk 7]
```diff
@@ -84,7 +84,7 @@ resource "project" "myproject" {
...
 - **member** (Block Set) Member of the project. Element has one to one mapping with the [JFrog Project Users API](https://www.jfrog.com/confluence/display/JFROG/Artifactory+REST+API#ArtifactoryRESTAPI-UpdateUserinProject). (see [below for nested schema](#nestedblock--member))
-- **repos** (Set of String) List of existing repo keys to be assigned to the project.
+- **repos** (Set of String) List of existing repo keys to be assigned to the project. By default, you can set a maximum of 100 repo keys. (see [below for limitations on the maximum number of repos](#repoNumLimitations)). You can override the system's default max cap by setting an environment value `REPO_LIMIT_OVERRIDE` to `true`. This setting will remove the default restriction of maximum allowable elements in the `repos` attribute. The default value of `REPO_LIMIT_OVERRIDE` is `false`. E.g. To override the default maximum allowable elements of `repos` attribute use `export REPO_LIMIT_OVERRIDE=true`.
```
Comment (alexhung) [inline docs/resources/project.md:87]: I think this is clearer:

> By default, `repos` is capped at 100 keys. (see [below for limitations on the maximum number of repos](#repoNumLimitations)). You can remove the system's cap by setting an environment value `REPO_LIMIT_OVERRIDE` to `true`. This setting will remove the restriction of maximum allowable elements in the `repos` attribute. The default value of `REPO_LIMIT_OVERRIDE` is `false`, e.g. To override, use `export REPO_LIMIT_OVERRIDE=true` in the shell.

File: pkg/project/resource_project.go [hunk 8]
```diff
@@ -217,9 +217,18 @@ func projectResource() *schema.Resource {
...
-			Elem:        &schema.Schema{Type: schema.TypeString},
+			Type:     schema.TypeSet,
+			Optional: true,
+			Elem: &schema.Schema{
+				Type: schema.TypeString,
+			},
+			MinItems: 0,
+			MaxItems: func() int {
+				if isOverride := getBoolEnvVar("REPO_LIMIT_OVERRIDE", false); isOverride {
+					return 1<<(32-1) - 1
```
Comment (alexhung) [inline pkg/project/resource_project.go:228]: This is not readable code:
1. this is not obvious what the return value is
2. why is this necessary? A simple integer suffice, no?

**Review by alexhung — CHANGES_REQUESTED** [review]

File: docs/resources/project.md [hunk 9]
```diff
@@ -128,3 +128,8 @@ Required:
 Optional:
 
 - **description** (String)
+
+## Note
+<a name="repoNumLimitations"></a>
+### Limitations of Repository assign/unassign to/from Project
+The project provider has limitations on its ability to assign or unassign a large number of repositories through the `repos` attribute. With more than a certain number of repos you might observe system disruptions and all the repositories might not be assigned/unassigned to/from the desired project. As per our analysis, it is **recommended to limit the number of repos to 100** to avoid internal platform limitations. The JFrog Engineering team is working on further improvement to ensure a better user experience.
```
[the remark this answers is not in the document]
Comment (dasmanas) [reply docs/resources/project.md:135]: Are suggesting a title change here? :)

File: pkg/project/resource_project_repo_test.go (hunk 4 above)
Comment (dasmanas) [reply pkg/project/resource_project_repo_test.go]: renamed variable

File: docs/resources/project.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (alexhung) [reply docs/resources/project.md:135]: I think @chb0github is referring to:

> The project provider has limitations

File: pkg/project/resource_project.go (hunk 8 above)
Comment (dasmanas) [reply pkg/project/resource_project.go:228]: True. Just mimic the equivalent of Max Int. I can set to a high value like 9999 instead.

File: docs/resources/project.md (hunk 7 above)
Comment (dasmanas) [reply docs/resources/project.md]: updated

File: pkg/project/resource_project.go (hunk 8 above)
Comment (dasmanas) [reply pkg/project/resource_project.go:228]: changed to 2147483647

File: docs/resources/project.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (dasmanas) [reply docs/resources/project.md]: verbiage updated

**Review by alexhung — DISMISSED** [review]

File: CHANGELOG.md [hunk 10]
```diff
@@ -1,3 +1,9 @@
+## 1.0.5 (Apr 05, 2022)
+
+IMPROVEMENTS:
```
Comment (maheshjfrog) [inline CHANGELOG.md:3]: Wondering if this PR is patch version change or a minor version change. Looks like not just a documentation update like mentioned in the description.

File: pkg/project/resource_project.go [hunk 11]
```diff
@@ -217,9 +217,18 @@ func projectResource() *schema.Resource {
...
+				Type: schema.TypeString,
+			},
+			MinItems: 0,
+			MaxItems: func() int {
+				if isOverride := getBoolEnvVar("REPO_LIMIT_OVERRIDE", false); isOverride {
+					return 2147483647
+				}
+				return 100
+			}(),
 			Description: "List of existing repo keys to be assigned to the project.",
```
Comment (maheshjfrog) [inline pkg/project/resource_project.go:232]: Can we add "Optional" and the default behavior in the Description?

**Review by maheshjfrog — COMMENTED** [review]
Minor comments on the version change and some description update. Rest are good.

File: CHANGELOG.md (hunk 10 above)
Comment (dasmanas) [reply CHANGELOG.md:3]: it is more than documentation change.

**request retry mechanism introduced on specific error conditions**

**Review by alexhung — APPROVED** [review]

**Review by danielmkn — APPROVED** [review]
